src/python/pants/option/custom_types.py

```diff
@@ -100,6 +100,37 @@ def shell_str(s: str) -> str:
     return s
 
 
+def memory_size(s: str) -> int:
+    """A string that normalizes the suffixes {GiB, MiB, KiB, B} into the number of bytes (int).
+
+    :API: public
+    """
+
+    def convert_to_bytes(power_of_2) -> int:
+        return int(float(s[:-3]) * (2 ** power_of_2))
+
+    original = s
+    s = s.lower().strip()
+
+    try:
+        return int(float(s))
+    except ValueError:
+        pass
+
+    if s.endswith("gib"):
+        return convert_to_bytes(30)
+    elif s.endswith("mib"):
+        return convert_to_bytes(20)
+    elif s.endswith("kib"):
+        return convert_to_bytes(10)
+    elif s.endswith("b"):
+        return int(float(s[:-1]))
```

Comment:
These units are sort of limited compared to what systems like Kubernetes support. As a user, I would wonder why I cannot just use a notation like "1.5G". 

See, for example, the [Kubernetes "Quantity" definition](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/#Quantity) and the [corresponding parser implementation](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go).

Reply:
Also, that example parser takes the position that floating point should be avoided for bytes quantities in order to maintain precision. (Thus it implements a fixed-point representation.)

Reply:
Seems easy enough to support those formats,no?

Reply:
Although, the thing about supporting `G` or `GB` as well as `GiB` is that it's confusing (a casual user might assume that `G` means binary gigabytes), and the decimal version is probably a lot less useful (who allocates RAM in decimal?)

Reply:
Exactly, that's why I went with GiB instead of G. No risk for ambiguity. 

Also, I don't want to over-engineer this. It's rare users would need to change this - we are not Terraform.

Reply:
Thoughts if we should accept floats like 1.5GiB? I don't want to do the full logic of Terraform..if we allow floats, there will be a little lossiness from round off error (which is fine imo)

If that lossiness concerns us, we should ban floats and require you to use a smaller unit, e.g. 0.5GiB becomes 512MiB.

Reply:
Other projects accept such numbers, would our users expect to be able to do so here as well?

Is there a Python module that already does these types of conversions? We are not the first project that wants to have units for bytes.

Reply:
So, this thing is very slow. I bisected to here and then added some timings around `import pint` and out pint usage. It turns out constructing the UnitRegistry takes ~175ms on my machine and we do this 10x. Even 1x I think 175ms is too much to sacrifice:
```
$ time ./pants -ldebug --no-anonymous-telemetry-enabled -V 2>&1 | grep UnitRegistry
>>> load pint.UnitRegistry: 0.16161394119262695s
>>> load pint.UnitRegistry: 0.16377687454223633s
>>> load pint.UnitRegistry: 0.16344141960144043s
>>> load pint.UnitRegistry: 0.16744184494018555s
>>> load pint.UnitRegistry: 0.16147994995117188s
>>> load pint.UnitRegistry: 0.16321659088134766s
>>> load pint.UnitRegistry: 0.19754433631896973s
>>> load pint.UnitRegistry: 0.16295242309570312s
>>> load pint.UnitRegistry: 0.1755695343017578s
>>> load pint.UnitRegistry: 0.16306495666503906s

real	0m2.447s
user	0m1.008s
sys	0m0.050s
```